### ISSUE TITLE
const introduction for universal quantification only on the RHS

### ISSUE THREAD

**Issue by roboguy13** [ku-fpg/hermit#134]
Is there a way to introduce a `const` call? For instance, transforming `x` into something like `const x b` for some `b`.

I'm trying to implement the transformation of the smallest free number function from Richard Bird's *Pearls of Functional Algorithm Design* and one of the steps (top of page 4) involves a law that has a universally quantified variable on the RHS but not the LHS. GHC doesn't like that if I try to put it into a RULE directly, so I'm trying to see if I can include it on the LHS with `const` and then use a `const` introduction in the HERMIT script.

The particular law is:

For any natural number `b`,

    [0..] \\ xs = let (us, vs) = partition (< b) xs
                   in ([0 .. b-1] \\ us) ++ ([b..] \\ vs)
(changed `where` to `let`)

The modified version that I'm considering with the `const` is:

    const ([0..] \\ xs) b = let (us, vs) = partition (< b) xs
                             in ([0 .. b-1] \\ us) ++ ([b..] \\ vs)

I'm thinking that this could be reasonable since a lambda is pretty analogous to a forall and we are using

    const a = \b -> a

**Comment by xich** [comment]
I've done something like this before to get around the need to have a function at the head of the rule, and to deal with variables that only appear on one side. Why not state your rule with the `const` (as you did above), then load it in as a lemma and use `modify-lemma` to just unfold `const`. This will give you the original lemma you desired, which you can then start proving.

**Comment by roboguy13** [comment]
@xich That works, but I'm running into a new problem. When I try to apply it, it says that the LHS doesn't match and I think this is because there is no universally quantified `b` where I am trying to apply the lemma (both the LHS of the lemma and what I'm trying to apply it are `\\ (enumFrom ▲ $fEnumInt (I# 0)) xs`).

Is there a way to introduce a forall in HERMIT?

**Comment by xich** [comment]
Yes, abstract-forall I think. I use it in the Century case study.

**Comment by roboguy13** [comment]
That sounds like what I'm looking for, but it looks like it expects a variable that is already being used. I want to introduce a new variable (I think).

So I have

    ∀ xs. minfreeFast xs ≡ head ▲ (\\ (enumFrom ▲ $fEnumInt (I# 0)) xs)

and I believe I would like to weaken this to get

    ∀ xs b. minfreeFast xs ≡ head ▲ (\\ (enumFrom ▲ $fEnumInt (I# 0)) xs)

The lemma (with the modifications applied) looks like this:

```
diff-partition (Assumed)
  ∀ xs b.
  \\ (enumFrom ▲ $fEnumInt (I# 0)) xs
  ≡
  (++) ▲
       (\\ (enumFromTo ▲ $fEnumInt (I# 0) ((-) ▲ $fNumInt b (I# 1)))
           (case partition ▲ (λ ds → (<) ▲ $fOrdInt ds b) xs of wild ▲
              (,) us vs → us))
       (\\ (enumFrom ▲ $fEnumInt b)
           (case partition ▲ (λ ds → (<) ▲ $fOrdInt ds b) xs of wild ▲
              (,) us vs → vs))
```

**Comment by xich** [comment]
Hmmm, let me make sure I'm understanding:

You have this rule:

```haskell
{-# RULES "blah" [~] forall xs b. const ([0..] \\ xs) b = let (us, vs) = partition (< b) xs in ([0 .. b-1] \\ us) ++ ([b..] \\ vs) #-}
```

Then you:

```
rule-to-lemma blah
modify-lemma blah (any-td (unfold 'const))
```

Now you should have:

```
forall xs b. [0..] \\ xs = let (us, vs) = partition (< b) xs in ([0 .. b-1] \\ us) ++ ([b..] \\ vs)
```

correct?

It is true that you will only be able to apply this lemma right-to-left, because `b` does not appear on the LHS. Even if it is universally quantified (which it still is), there is no way to invent an expression for it when instantiating the RHS with the expressions from the LHS pattern match. (You can still apply it right-to-left, because whatever expression `b` is can be thrown away... if that doesn't work it's a bug.)

I'm not sure how to really get around that. Do you need to apply it left-to-right to prove something else?

`abstract-forall` replaces all occurrences of an expression in the lemma with a new quantifier, so you are correct that it can't be used to just introduce a new quantifier, though I'm not sure if that would be useful... you would at least have to specify a type, but then it wouldn't bind over anything anyway.

**Comment by roboguy13** [comment]
Yeah, that looks correct.

I need to use it left-to-right. Let me give some broader context. Using that lemma, together with

```
head (xs ++ ys) = if null xs then head ys else head xs
```

we can show that

```
minfree :: [Nat] -> Nat
minfree xs = head ([0..] \\ xs)
```

can be transformed to

```
minfree xs =
  if null ([0..b-1] \\ us)
    then head ([b..] \\ vs)
    else head ([0..] \\ us)
  where (us, vs) = partition (< b) xs
```

which is the same as the original for any natural number chosen for `b`. Next, Bird transforms that to this (given the specification `minfrom a xs = head ([a..] \\ xs)`:

```
minfree xs = minfrom 0 xs

minfrom :: Nat -> [Nat] -> Nat
minfrom a xs
  | null xs            = a
  | length us == b - a = minfrom b vs
  | otherwise          = minfrom a us
    where (us, vs) = partition (< b) xs
```

After that, `b` is chosen to be ``a + 1 + (n `div` 2)``. Finally, an optimization is applied that avoids unnecessarily recalculating the length.

Maybe I'm going about this the wrong way.

**Comment by xich** [comment]
Ah, now I see why you would want to introduce the `const`! There is no way to rewrite just `[0..] \\ xs` because `b` would be unbound (as it is in your code snippet). So you need to rewrite `const ([0..] \\ xs) b` instead, right?

Could you do this:

State the lemma as:

```haskell
{-# RULES "blah" [~] forall xs b. const ([0..] \\ xs) b = let (us, vs) = partition (< b) xs in ([0 .. b-1] \\ us) ++ ([b..] \\ vs) #-}
```

Then prove it:

```
rule-to-lemma blah
prove-lemma blah
any-td (unfold 'const)
...
end-proof
```

Define an alternative `minfree'` which has a different type that accommodates the `b`:

```
minfree' :: [Nat] -> Nat -> Nat
minfree' xs b = head (const ([0..] \\ xs) b)
```

... do your derivation on `minfree'` as Bird does it.

Then, at the end, prove the following lemma to establish the connection between `minfree` and `minfree'`:

```
forall xs b. minfree xs = minfree' xs b
```

(the proof of which should just be an unfolding of `const`).

**Comment by roboguy13** [comment]
I had to swap the order of that last lemma so that `minfree'` is on the LHS so that GHC would accept it (since there needs to be a `b` on the LHS). I figure that I can use `lemmaBackward` eventually.

The problem now is that I'm not sure how to instantiate `b` to the specific value. If I just go into the `minfree` call from `main` and try to use `lemmaBackward` to replace it with `minfree'` HERMIT says `expression did not match the right-hand side` (which seems reasonable since it wouldn't know what `b` to use).

I'm thinking that I should make a function to choose the initial value for `b`

```
chooseB :: [Nat] -> Nat
chooseB xs = a + 1 + (n `div` 2)
  where
    a = 0
    n = length xs
```

but I'm not sure how to tell HERMIT to use this for `b`.

**Comment by xich** [comment]
Ah, once you prove `minfree` and `minfree'` equivalent, can you:

```
inst-lemma whatever-its-called 'b [| 1 + div (length xs) 2 |]
```

then use the instantiated version as the rewrite?

**Comment by roboguy13** [comment]
It looks like that runs into a HERMIT bug of some kind. I reduced it down to this (actually, you don't need to prove the lemma in order for this to happen, but you can):

```
hermit<0> rule-to-lemma minfree-const-eqv
minfree-const-eqv (Not Proven)
  ∀ xs b. minfree' xs b ≡ minfree xs
hermit<1> prove-lemma minfree-const-eqv ; assume
Goal:
  ∀ xs b. minfree' xs b ≡ minfree xs
Successfully proven: minfree-const-eqv
module Main where
  \\ ∷ [Nat] → [Nat] → [Nat]
  minfree ∷ [Nat] → Nat
  instB ∷ [Nat] → Nat
  minfrom ∷ Nat → (Nat, [Nat]) → Nat
  minfreeFast ∷ [Nat] → Nat
  const ∷ ∀ a b . a → b → a
  minfree' ∷ [Nat] → Nat → Nat
  descFrom ∷ Nat → [Nat]
  main ∷ IO ()
  main ∷ IO ()
hermit<3> inst-lemma minfree-const-eqv 'b [| 1 + div (length xs) 2 |]
ghc: panic! (the 'impossible' happened)
  (GHC version 7.10.1 for x86_64-unknown-linux):
	applyTypeToArgs
  Expression: I# 1 + div (length xs_a12x) (I# 2)
  Type: Int# -> Int
  Args: [1, +, div, length xs_a12x, I# 2]

Please report this as a GHC bug:  http://www.haskell.org/ghc/reportabug
```

**Comment by xich** [comment]
Yeah, the `+` needs a type and dictionary argument (because it is a type class method), and these aren't included in the expression (the parser doesn't infer them). This is why it'll probably be necessary to define your `chooseB` function and instantiate to that.

I've got some half-finished work on parsing Haskell expressions on a branch somewhere. I'll try to get that working at some point in the near future.

**Comment by xich** [comment]
Oh, and that actual error is because the parser doesn't parse infix operators, so it is trying to apply `I#` to all those arguments, which is a type mismatch.

**Comment by roboguy13** [comment]
Okay, so I ended up writing `chooseB` and using `inst-lemma minfree-const-eqv 'b [| chooseB xs |]` which seems to work fine. When I show the lemma, I get

```
minfree-const-eqv (Proven)
  ∀ xs. minfree' xs (chooseB xs) ≡ minfree xs
```

which looks good.

The problem is, when I go into `main` and try to use `lemmaBackward` on the call to `minfree`, it gives the same error as before:

```
hermit<3> rhs-of 'main
print ▲ $fShowInt (minfree (descFrom (I# 1000)))
hermit<4> app-arg
minfree (descFrom (I# 1000))
hermit<5> lemma-backward minfree-const-eqv
Error in expression: lemma-backward minfree-const-eqv
Rewrite failed:expression did not match right-hand side
```

For reference, `main` looks like

```
main :: IO ()
main = print (minfree (descFrom 1000))

descFrom :: Nat -> [Nat]
descFrom n = [n,n-1..0]
```

**Comment by xich** [comment]
Hmmm... out of curiosity, does it work if you do a `top ; flatten-module` before the `rhs-of 'main`?

I'm wondering if `minfree'`, `chooseB`, or `descFrom` in not in scope somehow. I don't think the error messages on the fold rewrite (which underlies the lemma rewrites) are too good.

If that doesn't work, can you check in what you have? I can tinker and try to figure out what is going wrong.

**Comment by roboguy13** [comment]
Actually, `flatten-module` seemed to fix that. Now I'm having some issues with using `diff-partition` because I can't seem to figure out how to instantiate the `Enum` dictionary that it expects to by the `Int` `Enum` dictionary.

I'm actually writing this in HERMIT-shell (although I tested all the stuff I said earlier in both just in case). I pushed `examples/minfree` to the `more-examples` branch of `ku-fpg/hermit-shell`.

Unfortunately, hermit-shell is a bit broken right now, so I would suggest not reinstalling it after you pull for the time being.

**Comment by xich** [comment]
Ah, glad to know `flatten-module` fixed that problem. For background, the top-level definitions are, by default, separated into strongly-connected components based on scoping. A given group can only see its own definitions and the definitions of the groups before it. `flatten-module` puts everything in a single group, restoring the behavior we expect (that all top-level things are in-scope throughout the module). Once we hand the program back to GHC, it will run SCC analysis and separate things again.

Anyway, I'll open a ticket so I remember to improve the error messages to reflect this.

As for the `Enum` dictionary... are you needing to instantiate one of the quantifiers of the lemma? Or are you trying to specialize some function to that dictionary?

**Comment by roboguy13** [comment]
For the `Enum` dictionary thing, I was accidentally bringing the `diff-partition` lemma back into scope at the shell prompt, so it overwrote my `inst-dictionaries` call.

Just now, I also removed the lemma modifications to `diff-partition` since they shouldn't be needed. Right now, I can't seem to convince HERMIT to let me use `diff-partition` (it gives a `expression did not match left-hand side` error).

It confuses me since the thing that I want to apply it to looks like this:

```
const ▲ ▲
      (head ▲ (\\ (enumFrom ▲ $fEnumInt (I# 0)) (descFrom (I# 1000))))
      (chooseB (descFrom (I# 1000)))
```

and the LHS of `diff-partition` is `const ▲ ▲ (\\ (enumFrom ▲ $fEnumInt (I# 0)) xs) b`.

**Comment by roboguy13** [comment]
Oops, I just realized that I had the `const` in the wrong spot. I moved it to where it should be, but I'm getting the same error.

```
HERMIT> appArg
head ▲
     (const ▲ ▲ (\\ (enumFrom ▲ $fEnumInt (I# 0)) (descFrom (I# 1000)))
            (chooseB (descFrom (I# 1000))))
[Done]

HERMIT> appArg
const ▲ ▲ (\\ (enumFrom ▲ $fEnumInt (I# 0)) (descFrom (I# 1000)))
      (chooseB (descFrom (I# 1000)))
[Done]

HERMIT> lemmaForward "diff-partition"
Rewrite failed:expression did not match left-hand side
*** Exception: failed to parse result value for rewrite (lemmaForward diff-partition): Null : Object not returned from Server
HERMIT> eval "show-lemma diff-partition"
diff-partition (Assumed)
  ∀ xs b.
  const ▲ ▲ (\\ (enumFrom ▲ $fEnumInt (I# 0)) xs) b
  ≡
  let ds =
        let ds = partition ▲ (λ ds → (<) ▲ $fOrdInt ds b) xs
            vs =
              case ds of wild ▲
                (,) us vs → vs
            us =
              case ds of wild ▲
                (,) us vs → us
        in (,) ▲ ▲ us vs
  in (++) ▲
          (\\ (enumFromTo ▲ $fEnumInt (I# 0) ((-) ▲ $fNumInt b (I# 1)))
              (case ds of ds ▲
                 (,) us vs → us))
          (\\ (enumFrom ▲ $fEnumInt b)
              (case ds of ds ▲
                 (,) us vs → vs))
[Done]
```

**Comment by roboguy13** [comment]
I'm not totally sure why this was a problem, but it got fixed when I removed the type annotations I had in the `diff-partition` RULE. I just pushed that change.
